### PR TITLE
ZCS-1088 : SIEVE: Any space(s) in header name should error out during…

### PR DESCRIPTION
**Pull request by chinmay15**
Problem : space in header name should not be allowed

Fix : added validation for header name in addheader, replaceheader, deleteheader, headerTest, addressTest, envelopTest, notifyMailto, enotify

Testing done: unit tests added for all above mentioned extensions

Testing to be done by QA: All above mentioned extensions should be tested by adding space in header name, which should fail.

### CODE REVIEW

File: store/src/java-test/com/zimbra/cs/filter/AddHeaderTest.java [hunk 1]
```diff
@@ -499,9 +499,16 @@ public void testAddHeaderNameAsMultipleSpace() {
     @Test
     public void testAddHeaderNameStartingWithSpace() {
         try {
-           String filterScript = "require [\"editheader\"];\n"
-                    + " addheader \" X-My-Test\" \"my-new-header-value\" \r\n"
-                    + "  ;\n";
+//           String filterScript = "require [\"editheader\"];\n"
+//                    + " addheader \" X-My-Test\" \"my-new-header-value\" \r\n"
+//                    + "  ;\n";
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/AddHeaderTest.java]: Remove commented lines (or add some explanation the reason why we need to leave the old filterScript)

File: store/src/java-test/com/zimbra/cs/filter/HeaderTest.java [hunk 2]
```diff
@@ -323,4 +323,70 @@ public void testHeaderMatchWithItself() throws Exception {
...
+        } catch (Exception e) {
+            e.printStackTrace();
+            fail("No exception should be thrown");
+        }
+    }
+
+    @Test
+    public void ttt() throws Exception {
+        String script =
+                "if anyof (header :contains [\"X_Header\"] \"123\") { "
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/HeaderTest.java]: Please add "require" clause.

File: store/src/java-test/com/zimbra/cs/filter/HeaderTest.java [hunk 3]
```diff
@@ -323,4 +323,70 @@ public void testHeaderMatchWithItself() throws Exception {
             fail("No exception should be thrown");
         }
     }
+
+    @Test
+    public void testHeaderNamesWithSpaces() throws Exception {
+        String script =
+                "if header :matches \" X-Header1\" \"*\" {"
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/HeaderTest.java]: Please add "require" clause

File: store/src/java-test/com/zimbra/cs/filter/AddressTest.java [hunk 4]
```diff
@@ -352,4 +352,79 @@ public void compareEmptyStringWithAsciiNumeric() {
...
+
+    @Test
+    public void compareHeaderNameWithLeadingAndTrailingSpaces() {
+        try {
+            Account acct = Provisioning.getInstance().getAccount(MockProvisioning.DEFAULT_ACCOUNT_ID);
+            RuleManager.clearCachedRules(acct);
+            Mailbox mbox = MailboxManager.getInstance().getMailboxByAccount(acct);
+
+            String filterScript =
+                    "if address :is :comparator \"i;ascii-numeric\" \" To \" \"test1@zimbra.com\" {"
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/AddressTest.java]: Please add "require" clause

File: store/src/java-test/com/zimbra/cs/filter/AddressTest.java [hunk 5]
```diff
@@ -352,4 +352,79 @@ public void compareEmptyStringWithAsciiNumeric() {
...
+
+    @Test
+    public void compareHeaderNameWithTrailingSpaces() {
+        try {
+            Account acct = Provisioning.getInstance().getAccount(MockProvisioning.DEFAULT_ACCOUNT_ID);
+            RuleManager.clearCachedRules(acct);
+            Mailbox mbox = MailboxManager.getInstance().getMailboxByAccount(acct);
+
+            String filterScript =
+                    "if address :is :comparator \"i;ascii-numeric\" \"To \" \"test1@zimbra.com\" {"
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/AddressTest.java]: please add "require" clause

File: store/src/java-test/com/zimbra/cs/filter/AddressTest.java [hunk 6]
```diff
@@ -352,4 +352,79 @@ public void compareEmptyStringWithAsciiNumeric() {
...
+
+    @Test
+    public void compareHeaderNameWithLeadingSpaces() {
+        try {
+            Account acct = Provisioning.getInstance().getAccount(MockProvisioning.DEFAULT_ACCOUNT_ID);
+            RuleManager.clearCachedRules(acct);
+            Mailbox mbox = MailboxManager.getInstance().getMailboxByAccount(acct);
+
+            String filterScript =
+                    "if address :is :comparator \"i;ascii-numeric\" \" To\" \"test1@zimbra.com\" {"
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/AddressTest.java]: Please add "require" clause

File: store/src/java-test/com/zimbra/cs/filter/AddHeaderTest.java [hunk 7]
```diff
@@ -499,9 +499,16 @@ public void testAddHeaderNameAsMultipleSpace() {
...
+            String filterScript =
+                    "require [\"log\",\"fileinto\",\"tag\"];"
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/AddHeaderTest.java]: Please add "editheader" capability string to the "require" clause

File: store/src/java-test/com/zimbra/cs/filter/AddHeaderTest.java [hunk 8]
```diff
@@ -499,9 +499,16 @@ public void testAddHeaderNameAsMultipleSpace() {
     @Test
     public void testAddHeaderNameStartingWithSpace() {
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/AddHeaderTest.java]: It seems this case actually tests not the header **name** with space, but the header **value** with space (\"  Val13\").  How about update the test case name as well?

File: store/src/java-test/com/zimbra/cs/filter/NotifyMailtoTest.java [hunk 9]
```diff
@@ -213,6 +213,23 @@ public void setUp() throws Exception {
           + "  :from \"test1@zimbra.com\" "
           + "  \"mailto:test2@zimbra.com?body=notifybody&message-id=dummymessageid&date=dummydate\";\n";
 
+    String filterScriptWithSpaceInHeaderName =
+            "require [\"enotify\", \"variables\"];\n"
```
Comment (ykomiyam) [inline store/src/java-test/com/zimbra/cs/filter/NotifyMailtoTest.java]: Please add "enotify" capability to the require clause as sell.

Comment (chinmay15) [inline store/src/java-test/com/zimbra/cs/filter/NotifyMailtoTest.java]: already there

**Comment by chinmay15** [comment]
@rupalid @ykomiyam done all the changes.

**Comment by chinmay15** [comment]
Merged into feature/sieve

commit 153274369c9886f01bf0806171d282ab13728253
Author: Chinmay Pathak <chinmay.pathak@synacor.com>
Date:   Fri Apr 7 17:58:08 2017 +0530

    ZCS-1088 : SIEVE: Any space(s) in header name should error out during validation

    Resolving review comments

**Pull request closed without merging**